Improve the style of the continuation prompt (PS2)

mafredri opened the pull request:
I've always thought the PS2 looked a bit out-of-place in Pure and I wanted to improve it, so I gave it a whack.

This PR does two things:

* Stylize the PS2 prompt with colors and the Pure prompt symbol `❯` (also includes the exit status in the PS2 prompt)
* Prevent prompt updates when PS2 is displayed (it's not possible to update PS1 with `reset-prompt` at this point)

This is what it looks like now:

<img width="489" alt="screenshot 2017-05-07 15 08 24" src="https://cloud.githubusercontent.com/assets/147409/25780838/94706e3e-3337-11e7-807a-3616a5e11b28.png">

A variation I tried was to include `❯` in front as well:

<img width="639" alt="screenshot 2017-05-07 15 06 25" src="https://cloud.githubusercontent.com/assets/147409/25780843/a3b538c0-3337-11e7-9838-b7dcb2172824.png">

But while looking OK was just not quite there.

I also looked at a bunch of different arrows and tried them out but couldn't find a good fit, closest ones were `↳` (but it did not quite align in the middle of the text, thus looking misaligned) and `⥅`. Neither of which I felt added value.

### Considerations

* Do we need to consider how this looks when virtualenvs (#254) are activated if we were to show them in the same manner?
* We could limit the number of entries (`cmdand`, etc.) by using `%N_` where `N` is the max number of entries, but I can't think of a very good reason to do this, at least it has never been an issue for me

sindresorhus commented:
Do we have to display `cmdand`? And I would go for a different color for the prompt character when in the continuation prompt, to signify it's a continuation. Maybe gray or dim cyan.

mafredri commented:
> Do we have to display `cmdand`?

I don't see a way around it, I think it's useful/important information. E.g. tells you why you ended up on a continuation prompt (might not always be obvious). Personally I wanted to try replacing it it with e.g. `&&`, but as far as I understand the (zsh) source, we cannot read the value of `%_` that would be printed to the terminal since it's done on-the-fly during prompt rendering/expansion and the only time the "execution context" is correct, is then.

> And I would go for a different color for the prompt character when in the continuation prompt, to signify it's a continuation. Maybe gray or dim cyan.

My thought was that `cmdand` was enough of an indication. Keeping the behavior of PS1 with regard to color was a simple solution, that IMO looks good. Not sure what a good color would be, if we go with `242` (same as `cmdand`) then my feeling is it doesn't really look like a prompt indication anymore, `cyan` on the other hand doesn't really change much (for me) compared to magenta:

<img width="631" alt="screenshot 2017-05-07 18 49 53" src="https://cloud.githubusercontent.com/assets/147409/25782554/fc4e3428-3355-11e7-9f54-ac0aefa32b81.png">
<img width="631" alt="screenshot 2017-05-07 18 49 48" src="https://cloud.githubusercontent.com/assets/147409/25782555/ffe32652-3355-11e7-88e5-98ec08d22a1b.png">

mafredri commented:
Looks like there's one aspect I did not consider, when there isn't any content for `%_`, currently looks like this (with cyan mod):

<img width="320" alt="screenshot 2017-05-07 18 55 25" src="https://cloud.githubusercontent.com/assets/147409/25782605/d3109708-3356-11e7-93b1-cf862b2c98fe.png">

mafredri commented:
So given the above, I think we definitely need to change the prompt color for continuations, or alternatively, use a different symbol (e.g. `❮` or `❯❯`). I'll try to experiment with some colors.

(Ignore the magenta color):

<img width="713" alt="screenshot 2017-05-07 19 10 27" src="https://cloud.githubusercontent.com/assets/147409/25782711/dd1487c6-3358-11e7-9af5-84ebfa79996e.png">

sindresorhus commented:
> Personally I wanted to try replacing it it with e.g. &&

That's what I was thinking too. `cmdand` doesn't look very good.

> So given the above, I think we definitely need to change the prompt color for continuations, or alternatively, use a different symbol (e.g. ❮ or ❯❯).

Do we need a symbol at all? The colors difference kinda makes the distinction.

Alternatively:

```
❯ echo -n Hello &&
cmdand └ foo
```


```
❯ echo -n Hello &&
cmdand │ foo
```

mortenvn commented:
I'm really like the new style! It fits right in with the rest of Pure. 👍

#### Some thoughts on the things you've already discussed

> Personally I wanted to try replacing it it with e.g. &&

Shame that this isn't possible. It would have been glorious! 

It took me a while to figure out that `cmdand` actually meant `&&` and that it wasn't some sort of bug or placeholder text. (I'm not a smart man 😩)

> Do we need a symbol at all?

In my opinion, the symbol is a great way to indicate that everything fine, that your still in the Pure prompt and that you can keep writing commands.

> I think we definitely need to change the prompt color for continuations, or alternatively, use a different symbol (e.g. ❮ or ❯❯).

Of the ideas presented, I really liked `❯❯`. It's simple, it looks like the Pure symbol, but at the same time looks different enough to be noticeable.

#### Delimiter choices (to support virtualenv)
As we discussed in https://github.com/sindresorhus/pure/pull/254, I tried to see how this could be combined with virtualenv support:

![delimiter-type](https://cloud.githubusercontent.com/assets/3471625/25823005/f1403186-343a-11e7-89b5-c384db52de62.png)

As you can see I just smacked pipe symbols of different weights between them. I think nr. 2 (medium vertical bar) is the best one as the pipe has the same thickness as ❯.

### Color choices
![colors](https://cloud.githubusercontent.com/assets/3471625/25823368/32e70096-343c-11e7-9981-1e90dba21895.png)
I prefer the first two, with a slight edge towards nr. 1. This is really subjective, but I just think it looks the cleanest and most consistent.

There is, however, a case to be made for the cyan one as well: When you have text (like "cmdand") I don't think the color matters that much since it remains easy to understand that you've entered _"Fuck it! We'll do it live!"_ mode. But in the first example (`echo Hello \`) we don't have that luxury. In this case the cyan color might help indicate the change. 

I'm not a big fan of making the ❯ symbol grey. Having three grey things (virtualenv, the text and the ❯) next to each other makes it harder to parse. I also feel that the color helps guide the eye and indicate that the prompt is ready for more.

### Alternative solution
I don't really know how successful this proposal is, but I was curious to see what could be done without the text:
![alternative](https://cloud.githubusercontent.com/assets/3471625/25826645/dd30c018-3447-11e7-8d6f-44a56eb3a917.png)

In the end I think it's an interesting idea, but will probably cause some confusion. For example: Note how you go right from three prompts (❯❯❯) to being finished. Intuitively I would have expected to get back to two before it quit.

sunjon commented:
The last proposal(venv  + alt color) looks great. Is there something stopping it being able to drop down an indentation level on the first 'fi' ?

mafredri commented:
Thanks for the awesome feedback @mortenvn, you've given me a bit to think about, I'll try to get back to this in the coming days 😄.

> Shame that this isn't possible. It would have been glorious!

Yup 😢.

sindresorhus commented:
> Shame that this isn't possible. It would have been glorious!
> > Yup 😢.

Open a feature request on the ZSH issue tracker?

sindresorhus commented:
@mafredri How should we proceed with this?

mafredri commented:
@sindresorhus apologies for leaving this hanging, I'll try to get around to fixing this up during this week. I need to reacquaint myself with this again and consider @mortenvn suggestions 😄.

> Open a feature request on the ZSH issue tracker?

They don't really have an issue tracker, just the zsh-workers mailing list (and I dread mailing lists). Anywho, I'll try to get around to posting this on there, maybe we could have this in Zsh 5.4 if we're lucky.

mafredri commented:
Heading off to vacation now so I didn't get this quite done. I did a bit of experimenting and testing mortenvn's suggestions though, and I can't help but feel that the PS2 gets crowded, esp. when a virtualenv is active. An idea I had was to instead just show:

```
myvirtialenv ❯ echo hi &&
... andand ❯ echo bye
```

The `...` would indicate PS2 (continuation prompt).

Colors would be dark as above for everything before the `❯`.

sindresorhus commented:
@mafredri What should we do about this? Happy to go with the simpler solution.

mafredri commented:
I've rebased the branch and simplified the continuation prompt, it no longer displays virtualenv and uses `...` to signify that we have entered continuation mode.

<img width="201" alt="Screen Shot 2019-06-11 at 23 04 20" src="https://user-images.githubusercontent.com/147409/59303111-30070a80-8c9e-11e9-8951-463d78dca247.png">

<img width="243" alt="image" src="https://user-images.githubusercontent.com/147409/59303102-2aa9c000-8c9e-11e9-9c39-947b01a04730.png">

I think this works good enough to be merged, thoughts?
